File: Back-End/src/main/java/iuh/fit/ecommerce/services/CartService.java
package iuh.fit.ecommerce.services;

import iuh.fit.ecommerce.dtos.request.cart.CartAddRequest;
import iuh.fit.ecommerce.dtos.request.cart.CartUpdateQuantityRequest;
import iuh.fit.ecommerce.dtos.response.cart.CartResponse;
import iuh.fit.ecommerce.dtos.response.cart.CartWithCustomerResponse;
import org.springframework.data.domain.Page;

import java.util.List;

public interface CartService {
    CartResponse getOrCreateCart();

    CartResponse addProduct(CartAddRequest request);

    CartResponse removeProduct(Long productVariantId);

    void clearCart(Long userId);

    CartResponse updateProductQuantity(CartUpdateQuantityRequest request);

    // Admin methods
    Page<CartWithCustomerResponse> getAllCartsWithItems(int page, int size, String keyword);

    void sendRemindersBatch(List<Long> cartIds);

    CartWithCustomerResponse getCartByCustomerId(Long customerId);
}

File: Front-End/src/services/cart.service.ts
import axiosClient from "@/configurations/axios.config";
import type {
  CartAddRequest,
  CartResponse,
  CartWithCustomerResponse,
} from "@/types/cart.type";

export const cartService = {
  getCart: async () => {
    const response = await axiosClient.get<CartResponse>(`/carts`);
    return response.data;
  },
  addProductToCart: async (request: CartAddRequest) => {
    const response = await axiosClient.post<CartResponse>(
      `/carts/add`,
      request
    );
    return response.data;
  },

  removeProductFromCart: async (productVariantId: number) => {
    const response = await axiosClient.delete<CartResponse>(
      `/carts/remove/${productVariantId}`
    );
    return response.data;
  },

  updateCartItemQuantity: async (
    productVariantId: number,
    quantity: number
  ) => {
    const response = await axiosClient.put<CartResponse>(
      `/carts/update-quantity`,
      {
        productVariantId,
        quantity,
      }
    );
    return response.data;
  },

  clearCart: async (userId: number) => {
    const response = await axiosClient.delete(`/carts/clear/${userId}`);
    return response.data;
  },

  // Admin/Staff methods
  getAllCarts: async (
    page: number = 0,
    size: number = 10,
    keyword?: string
  ) => {
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
    });
    if (keyword) {
      params.append("keyword", keyword);
    }
    const response = await axiosClient.get<CartWithCustomerResponse>(
      `/carts/admin/all?${params.toString()}`
    );
    return response.data;
  },

  getCartByCustomerId: async (customerId: number) => {
    const response = await axiosClient.get(
      `/carts/admin/customer/${customerId}`
    );
    return response.data;
  },

  sendRemindersBatch: async (cartIds: number[]) => {
    const response = await axiosClient.post(
      `/carts/admin/send-reminders`,
      cartIds
    );
    return response.data;
  },
};
